xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp

```diff
@@ -541,8 +541,20 @@ void CGUIDialogPVRTimerSettings::OnSettingAction(const std::shared_ptr<const CSe
     m_startLocalTime.GetAsSystemTime(timerStartTime);
     if (CGUIDialogNumeric::ShowAndGetTime(timerStartTime, g_localizeStrings.Get(14066)))
     {
+      const auto timerSpan =
+          (m_endLocalTime > m_startLocalTime ? (m_endLocalTime - m_startLocalTime)
+                                             : CDateTimeSpan(0, 1, 0, 0));
```

Comment:
why is timerSpan calculated always if we need it only inside the following if block?

Reply:
is the new logic actually correct? shouldn't we compare against the **new** start time, not the old one?

Reply:
@ksooo : We want to know the old span for the timer BEFORE it was updated (using the dialog). The line below it updates m_endStartLocalTime there, so we need to calculate it before that. When calculating it, you want to make sure it's valid therefor we fallback to a span of an hour, in case it isn't. This is also what the TVHeadend-addon uses as default for timers without an end-time. I don't know how the other PVR addons behave, but this logic seems sane to me.

Reply:
May I ask what the actual problem you want to solve is? How can I reproduce the problem (using pvr.hts at best)?

Reply:
Sorry, with this instructions i run not in any error. The timer rule is saved just fine. `Validate` returns true.

Reply:
Voila is French for: "There it is (-> the error box that Validate() fails)". I think for you to reproduce this with my instructions, you probably also need to change the "Start time window calculation"-setting in the TVHeadend HTSP client addon to "Relaxed (start time +/- margin)".

Also note that the issue also occurs when editing already existing timer rules (that's where I first noticed the issue). There it's especially annoying.

Reply:
Patching end time for a timer not supporting end time seems a hack to me. The correct fix imo is:

```diff
diff --git a/xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp b/xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp
index dfb8d84f66..0f3bd35587 100644
--- a/xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp
+++ b/xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp
@@ -564,12 +564,14 @@ bool CGUIDialogPVRTimerSettings::Validate()
   bool bStartAnyTime = m_bStartAnyTime;
   bool bEndAnyTime = m_bEndAnyTime;
 
-  if (!m_timerType->SupportsStartAnyTime() ||
-      !m_timerType->IsEpgBased()) // Start anytime toggle is not displayed
+  if (m_timerType->SupportsStartTime() &&
+      (!m_timerType->SupportsStartAnyTime() ||
+       !m_timerType->IsEpgBased())) // Start anytime toggle is not displayed
     bStartAnyTime = false; // Assume start time change needs checking for
 
-  if (!m_timerType->SupportsEndAnyTime() ||
-      !m_timerType->IsEpgBased()) // End anytime toggle is not displayed
+  if (m_timerType->SupportsEndTime() &&
+      (!m_timerType->SupportsEndAnyTime() ||
+       !m_timerType->IsEpgBased())) // End anytime toggle is not displayed
     bEndAnyTime = false; // Assume end time change needs checking for
 
   // Begin and end time
```

This patch also fixes your issue @arnova.

Maybe @phunkyfish can check whether the use case that led to the introduction of `Validate()`still works with this patch?

Reply:
Also looks ok to me, but can't you simply drop IsEpgBased() then when adding SupportsStartTime()/SupportsEndTime() ?

I'm ok with both my and your solution but please note that the TVHeadend-addon also patches the endtime (starttime + 1 hour) for timer rules without an end-time.

Reply:
I can check it but it will be at least a week before I can so.

Reply:
>  but can't you simply drop IsEpgBased() then when adding SupportsStartTime()/SupportsEndTime() ?

Why?

Reply:
`Save()` calls `Validate()` so the logic is only required in one place.

Reply:
> When both !m_bStartAnyTime and !m_bEndAnyTime, check if endTime < startTime" ?

Almost. times must only be compared if both start and end time are supported by the timer type.

So, it could be like:

```c++
If (endtime_supported && !end_anytime && 
     starttime_supported && !start_anytime && 
     endtime < starttime)
{
  <error box>
  return false;
}
return true;
```

Reply:
> > When both !m_bStartAnyTime and !m_bEndAnyTime, check if endTime < startTime" ?
> 
> Almost. times must only be compared if both start and end time are supported by the timer type.
> 
> So, it could be like:
> 
> ```c++
> If (endtime_supported && !end_anytime && 
>     starttime_supported && !start_anytime && 
>     endtime < starttime)
> {
>   <error box>
>   return false;
> }
> return true;
> ```

@ksooo / @phunkyfish : I think I've fixed (and simplified) the logic, please check. I think the logic in Validate() was flawed for more cases than the one I stumbled on, when I looked at it more closely. I think the root cause was that, in IMHO, the logic in Validate() was almost unreadable.